draft-ietf-taps-impl.md

```diff
@@ -1030,39 +1048,78 @@ Data Unit: Message
+The following explanation assumes a client-server communication model.
+
+Stream mapping requires an association to already be in place between the client
+and the server, and it requires the server to understand that a new incoming stream
+should be represented as a new Connection Object by the Transport Services system.
+To signal the use of stream mapping, the server SHOULD use the adaptation layer indication
+parameter of LISTEN.SCTP and the client SHOULD use the adaptation layer indication
+parameter of CONNECT.SCTP, both with the "Stream Mapping" adaptation code point.
+Reception of an adaptation layer indication notification with the same adaptation code point value
+then indicates that the peer is capable of stream mapping.
+
+In the absence of an adaptation layer indication notification carrying a "Stream Mapping"
+adaptation code point, stream mapping can still be
+used if the application is aware that the peer application does not directly access SCTP streams.
+If, for example, a transport-protocol-agnostic client application tries to connect to a
+native SCTP server application that does not implement stream mapping, and assumes
+that a new Message sent on this Connection will look like the first Message of a new
+Connection on the server side, communication will fail because the
+SCTP-based application on the server side does not obtain a new Connection.
+As a counter-example, if both the client and server are
+written over a Transport Services system, and the server listens on all available protocols, stream
+mapping works without the application having to be aware of it (a code example that works
+without the adaptation layer notification, described in {{NEATPy}}, is available from the
+NEATPy repository mentioned in {{appendix-implementations}} at the time of writing).
+
+A new SCTP stream is created by sending an SCTP message with a new stream id. Thus, to
+implement stream mapping,
+the Transport Services system MUST provide a newly created Connection Object to the
+application upon the reception of such a message. The necessary semantics to
+implement a Transport Services system's `Close` and `Abort` primitives are provided
+by the stream reconfiguration (reset) procedure described in {{?RFC6525}}. This also
+allows to re-use a stream id after resetting ("closing") the stream.
+To implement this functionality, SCTP stream reconfiguration {{?RFC6525}} MUST be
+supported by both the client and the server side.
+
+To avoid head-of-line blocking, stream mapping SHOULD only be implemented
+when both sides support message interleaving {{?RFC8260}}. This allows a sender to
+schedule transmissions between multiple streams without risking that transmission
+of a large message on one stream might block transmissions on other streams
+for a long time.
+
+To avoid conflicts between stream ids, the following procedure is recommended:
```

Comment:
Please help someone who hasn't read the SCTP specs in 15 years: is there really no deconfliction when there is simultaneous open of a new stream ID?

Reply:
I wouldn't really know myself. This is how Felix Weinrank and Michael Tuexen implemented it in NEAT - so I figured that this must be necessary. I can check with them if there's really no other way... but if there were, they would probably have used it. This said, this is also a few years ago now.